src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -86,86 +139,138 @@ namespace NovelRT::Graphics
+            GetOrCreateGraphicsBufferForAllocationSize(sizeToStage, GraphicsBufferKind::Index);
+        auto indexRegion = indexBuffer->Allocate(sizeToStage, alignment);
+        auto writeArea = stagingBuffer->Map<uint8_t>(indexRegion);
+#ifdef WIN32
+        memcpy_s(writeArea, sizeToStage, data, sizeToStage);
+#else
+        memcpy(writeArea, data, sizeToStage);
+#endif
+        stagingBuffer->UnmapAndWrite(indexRegion);
+        currentContext->Copy(indexBuffer, stagingBuffer);
+
+        return indexRegion;
+    }
+
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetStagingBufferWithProperSizeHandling(
         size_t sizeToStage,
-        std::shared_ptr<GraphicsContext>& currentContext)
+        std::shared_ptr<GraphicsContext>& /*currentContext*/)
     {
-        std::shared_ptr<GraphicsBuffer> stagingBuffer = _stagingBuffer.getActual();
+        _stagingBufferSize = _stagingBufferSize < sizeToStage ? sizeToStage : _stagingBufferSize;
+
+        std::shared_ptr<GraphicsBuffer> chosenBuffer = nullptr;
+
+        for (auto&& buffer : _stagingBuffers[_contextIndex])
+        {
+            if (buffer->GetLargestFreeRegionSize() >= sizeToStage)
+            {
+                chosenBuffer = buffer;
+                break;
+            }
+        }
 
-        if (stagingBuffer->GetLargestFreeRegionSize() < sizeToStage)
+        if (chosenBuffer == nullptr)
         {
-            _stagingBufferSize = (_stagingBufferSize * 2) + sizeToStage;
-            std::shared_ptr<GraphicsBuffer> newStagingBuffer = CreateStagingBuffer();
-            currentContext->Copy(newStagingBuffer, stagingBuffer);
-            currentContext->EndFrame();
-            _graphicsDevice->Signal(currentContext->GetFence());
-            _graphicsDevice->WaitForIdle();
-            _stagingBuffer.reset(newStagingBuffer);
-            stagingBuffer = _stagingBuffer.getActual();
-            currentContext->BeginFrame();
+            chosenBuffer = CreateStagingBuffer();
         }
-        return stagingBuffer;
+
+        return chosenBuffer;
     }
 
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::CreateStagingBuffer()
     {
-        return _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
+        auto bufferToReturn = _graphicsDevice->GetMemoryAllocator()->CreateBufferWithDefaultArguments(
             Graphics::GraphicsBufferKind::Default, Graphics::GraphicsResourceAccess::Write,
             Graphics::GraphicsResourceAccess::Read, _stagingBufferSize);
+
+        _stagingBuffers[_contextIndex].emplace_back(bufferToReturn);
+
+        return bufferToReturn;
     }
 
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetOrCreateGraphicsBufferForAllocationSize(
-        size_t allocationSize)
+        size_t allocationSize,
+        GraphicsBufferKind bufferKind)
     {
-        for (auto&& buffer : _vertexBuffers)
+        std::vector<std::shared_ptr<GraphicsBuffer>>* buffers = nullptr;
```

Comment:
I'd prefer using a lambda here, something like `const auto buffers = [&]{ /* the switch statement, using return instead */ }();`.

Reply:
I don't know how this would look at all

Reply:
Something like this:
```cpp
const auto buffers = [&]{
    switch (bufferKind) {
        case GraphicsBufferKind::Default:
            return _defaultBuffers;
        // etc.
    }
    // fail-fast for invalid bufferKind
}();
````

Reply:
If it helps, the C++ core guidelines has a rule for this to help explain it: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init

Reply:
Ah, that's interesting, so its a const correctness thing then?

Reply:
Yeah, a little bit. It also just makes it clearer what values are allowed, and keeps the selection colocated with the variable so that it can't get easily shuffled around in a refactor by accident.